firmware/smolusb/src/descriptor/microsoft10.rs

```diff
@@ -0,0 +1,245 @@
+}
+
+// - StringDescriptor ---------------------------------------------------------
+
+/// Microsoft OS 1.0 String Descriptor Header
+#[derive(Clone, Copy)]
+pub struct StringDescriptor<'a> {
+    pub head: StringDescriptorHeader,
+    pub tail: [u8; 16],
+    _marker: PhantomData<&'a ()>,
+}
+
+impl<'a> StringDescriptor<'a> {
+    #[must_use]
+    #[allow(clippy::cast_possible_truncation)]
+    pub const fn new(vendor_code: u8) -> Self {
+        Self {
+            head: StringDescriptorHeader {
+                bLength: 18,
+                bDescriptorType: DescriptorType::String as u8,
+            },
+            #[rustfmt::skip]
+            tail: [
+                // Signature: "MSFT100"
+                b'M', 0x00, b'S', 0x00, b'F', 0x00, b'T', 0x00, b'1', 0x00, b'0', 0x00, b'0', 0x00,
+                // Vendor Code
+                vendor_code,
+                // Padding
+                0x0,
+            ],
+            _marker: PhantomData,
+        }
+    }
+}
+
+impl<'a> StringDescriptor<'a> {
+    /// Returns an iterator to the descriptor
+    #[allow(clippy::cloned_instead_of_copied)]
+    #[allow(clippy::iter_without_into_iter)]
+    pub fn iter(&'a self) -> StringDescriptorIterator<'a> {
+        let head_iter: slice::Iter<'a, u8> = self.head.as_iter();
+        let tail_iter: slice::Iter<'a, u8> = self.tail.iter();
+        head_iter.cloned().chain(tail_iter.cloned())
+    }
+}
+
+pub type StringDescriptorIterator<'a> =
+    iter::Chain<iter::Cloned<slice::Iter<'a, u8>>, iter::Cloned<slice::Iter<'a, u8>>>;
+
+// - CompatibleIdFeatureDescriptor --------------------------------------------
+
+/// Microsoft OS 1.0 Compatible ID Feature Descriptor header
+#[derive(AsBytes, FromBytes, FromZeroes, Clone, Copy)]
+#[repr(C, packed)]
+pub struct CompatibleIdFeatureDescriptorHeader {
+    pub dwLength: u32,
+    pub bcdVersion: u16, // 0x0100
+    pub wIndex: u16,     // 0x0004 Compatible ID Feature Descriptor
+    pub bCount: u8,
+    pub aReserved: [u8; 7],
+}
+
+impl CompatibleIdFeatureDescriptorHeader {
+    #[must_use]
+    pub const fn new() -> Self {
+        Self {
+            dwLength: 0,
+            bcdVersion: 0x0100, // v1.0
+            wIndex: 0x0004,     // Compatible ID Feature Descriptor
+            bCount: 0,
+            aReserved: [0; 7],
+        }
+    }
+}
+
+impl AsByteSliceIterator for CompatibleIdFeatureDescriptorHeader {}
+
+/// Microsoft OS 1.0 Compatible ID Feature Descriptor function section
+#[derive(AsBytes, FromBytes, FromZeroes, Clone, Copy)]
+#[repr(C, packed)]
+pub struct CompatibleIdFeatureDescriptorFunction {
+    pub bFirstInterfaceNumber: u8,
+    pub bReserved: u8,
+    pub aCompatibleId: [u8; 8],
+    pub aSubCompatibleId: [u8; 8],
+    pub aReserved: [u8; 6],
+}
+
+impl CompatibleIdFeatureDescriptorFunction {
+    #[must_use]
+    pub const fn new() -> Self {
+        Self {
+            bFirstInterfaceNumber: 0,
+            bReserved: 0x01, // TODO does this _need_ to be 0x01 ???
```

Comment:
It should be `0x01` according to the spec.

Reply:
Tx! 3826825febcae83d23998591ae862dcb7adffcf0